mesonbuild/dependencies/coarrays.py

```diff
@@ -21,15 +21,15 @@
 from .factory import factory_methods
 
 if T.TYPE_CHECKING:
-    from . base import DependencyType
+    from . base import DependencyType, Dependency
     from ..environment import Environment, MachineChoice
 
 
 @factory_methods({DependencyMethods.PKGCONFIG, DependencyMethods.CMAKE, DependencyMethods.SYSTEM})
 def coarray_factory(env: 'Environment', for_machine: 'MachineChoice',
-                    kwargs: T.Dict[str, T.Any], methods: T.List[DependencyMethods]) -> T.List['DependencyType']:
+                    kwargs: T.Dict[str, T.Any], methods: T.List[DependencyMethods]) -> T.List[T.Callable[[], 'Dependency']]:
     fcid = detect_compiler('coarray', env, for_machine, 'fortran').get_id()
-    candidates = []  # type: T.List[DependencyType]
+    candidates: T.List[T.Callable[[], 'Dependency']] = []
```

Comment:
I think this should be `ExternalDependency`

Reply:
It would make sense, but `factory.py` is also using `Dependency` instead of `ExternalDependency`. And changing it to `ExternalDependency` for just `coarrays.py` makes mypy unhappy :(

Since the plan of this PR series is to "just" got the dependencies fully annotated, I would say that we should refactor this after all PRs are merged.